### ISSUE TITLE
install from source with pip < 22.3 missing Python files

### ISSUE THREAD

**Issue by minrk** [h5py/h5py#2219]
**tl;dr:** if folks are seeing `AttributeError: h5py has no attribute File` or some such, h5py 3.8 installed from source with an older pip results in an incomplete installation. Upgrading pip `pip install --upgrade pip` to at least 22.3 fixes the issue.

## Description

```
python3 -m pip install --no-binary h5py h5py
```

is successfully building the extensions, but not installing any Python files, when running with `pip<22.3` (no .py files installed at all). 

To assist reproducing bugs, please include the following:

 * Operating System ubuntu 22.04 (docker: `dolfinx/dev-env:v0.6.0`)
 * Python version 3.10
 * Where Python was acquired (apt, via docker image: `dolfinx/dev-env:v0.6.0`)
 * h5py version 3.8.0
 * HDF5 version 1.12.2
 * The full traceback/stack trace shown (if it appears)

```
python -c 'import h5py; print(h5py.version.info)'
```

produces:

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: module 'h5py' has no attribute 'version'
``` 

because `h5py/__init__.py` is not installed.

## repro steps

Start container:

```bash
docker run --rm -it dolfinx/dev-env:v0.6.0 bash
```

in container:

```bash
python3 -m pip install 'pip==22.2.*'
python3 -m pip install --no-binary h5py h5py
```

Then test:

```
root@0336adf39396:~# python3 -c 'import h5py; print(h5py.version.info)'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AttributeError: module 'h5py' has no attribute 'version'
```

`ls` shows that none of the Python files are installed:

```
# ls /usr/local/lib/python3.10/dist-packages/h5py
_conv.cpython-310-aarch64-linux-gnu.so     _proxy.cpython-310-aarch64-linux-gnu.so     h5.cpython-310-aarch64-linux-gnu.so    h5d.cpython-310-aarch64-linux-gnu.so   h5fd.cpython-310-aarch64-linux-gnu.so  h5l.cpython-310-aarch64-linux-gnu.so  h5pl.cpython-310-aarch64-linux-gnu.so  h5t.cpython-310-aarch64-linux-gnu.so
_errors.cpython-310-aarch64-linux-gnu.so   _selector.cpython-310-aarch64-linux-gnu.so  h5a.cpython-310-aarch64-linux-gnu.so   h5ds.cpython-310-aarch64-linux-gnu.so  h5g.cpython-310-aarch64-linux-gnu.so   h5o.cpython-310-aarch64-linux-gnu.so  h5r.cpython-310-aarch64-linux-gnu.so   h5z.cpython-310-aarch64-linux-gnu.so
_objects.cpython-310-aarch64-linux-gnu.so  defs.cpython-310-aarch64-linux-gnu.so       h5ac.cpython-310-aarch64-linux-gnu.so  h5f.cpython-310-aarch64-linux-gnu.so   h5i.cpython-310-aarch64-linux-gnu.so   h5p.cpython-310-aarch64-linux-gnu.so  h5s.cpython-310-aarch64-linux-gnu.so   utils.cpython-310-aarch64-linux-gnu.so
```

Upgrading pip to 22.3.1 fixes this:

```
python3 -m pip uninstall h5py
python3 -m pip install 'pip==22.3.*'
python3 -m pip install -vv --no-binary h5py h5py
python3 -c 'import h5py; print(h5py.version.info)'
```

gives:

```
Summary of the h5py configuration
---------------------------------

h5py    3.8.0
HDF5    1.12.2
Python  3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.23.3
cython (built with) 0.29.33
numpy (built against) 1.21.6
HDF5 (built against) 1.12.2

```

Weirdly, the versions of setuptools and friends inside the build environment don't seem to matter (they are the same in both), it's the _host_ pip that's requred to be up-to-date to properly install the package files. I don't really understand that or see [what change is relevant](https://pip.pypa.io/en/latest/news/#v22-3), or what the right workaround (presumably in pyproject.toml and/or setup.py) is.

[full debug session output](https://gist.github.com/minrk/7295212e77d712932dbfde84215b2e09)

**Comment by minrk** [comment]
On further debugging, since this is a debian system Python, setting `DEB_PYTHON_INSTALL_LAYOUT=deb_system` [per setuptools discussion](https://github.com/pypa/setuptools/issues/3269#issuecomment-1254507377) also fixes the issue without needing to upgrade pip.

If I've understood correctly (I'm not sure I have), I think this can probably safely be considered a fairly specific pip+setuptools+debian problem that doesn't need to be fixed (if it even can be) at the package level.
